### PR TITLE
pivot.arp_scan()

### LINKED ISSUE #239: Implement pivot.arp_scan
**Is your feature request related to a problem? Please describe.**
Currently the only way for eldritch to discover hosts is through the `pivot.port_scan` function. This can be time consuming and noisy. To provide other stealthier option we should add a function that queries the systems arp table similar to `$ arp -a`.

**Describe the solution you'd like**
Implement a function that returns a list of type `Dict`.
The returned dictionaries should have the structure:
```json
{
    "ip":"192.168.119.21",
    "mac":"98:e7:43:3f:0a:28",
    "interface":"eth0",
}
```
The `arp-scan-rs` project seems like a promising way to go.
It may even give us the vendor OU for free.
https://github.com/kongbytes/arp-scan-rs

**Describe alternatives you've considered**
N/a

**Additional context**
N/a

### PR DESCRIPTION
**Pull request by jabbate19**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Adds ARP scan capability.

Will scan given subnets' IPs using ARP requests and checking for ARP replies. Does this multithreaded, one thread per subnet.

#### Which issue(s) this PR fixes:
Fixes #239 

### Note:
I will take any advice on how to test this!

### CODE REVIEW

File: implants/lib/eldritch/src/pivot.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot.rs:92]: May wait on this want to figure out  the cargo clippy PR before merging any new formatted code.

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:177]: Please use lower case variables it’s easier to remember for users if every variable is lower case.

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:173]: Good job using dictionaries 🥳 they’re not the most intuitive.

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:156]: Please avoid using unwrap in production code.
It’s better to either bubble up with a question mark or in this case since we’re in the root function handle the error returning a verbose anyhow error like “apr_scan: failed to”

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:85]: Unwrap

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:78]: Unwrap

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:47]: Unwrap

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:62]: Unwrap

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:128]: Unwrap

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:131]: I have a medium preference towards returning the original error string at the end of the anyhow error but I’ll defer to you if that makes sense Eg if no other error can return may not make sense.

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:166]: Unwrap

**Review by hulto — CHANGES_REQUESTED** [review]
🌶️ feature very excited to have this.
Few change requests.

**Comment by hulto** [comment]
Also please add tests and docs.
Test could be as simple as run it through the interpreter passing a fresh heap make sure it runs successfully and sees a MAC address.

there may also be a few unwraps I didn’t tag.

File: implants/lib/eldritch/Cargo.toml
Comment (hulto) [inline implants/lib/eldritch/Cargo.toml:44]: Please add these to the workspace root and use workspace true in the package. This way if another package wants to use a crate they'll be kept in step with the main version. If specific libs need to break with the current workspace version we can do that but should avoid it.

File: implants/lib/eldritch/src/pivot.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot.rs:86]: Can we leave this as a single line for now. Would like to maintain consistency even while we look to auto formatting.

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:24]: Thank you for making a struct 🙏

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:36]: Would it make sense to return an error? Might be nice in case something breaks to know why.

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:44]: I believe if you return an error here instead of panicking the thread will close and pass back the error message to join similar to panic.
If we can do that instead that would be my preference even if it only achieves the same effect as panic it will be more reusable and also clearer that the function itself shouldn't panic.

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:112]: We want to avoid println in production code eldritch should only print when a user calls print (print calls in eldritch have a special handler to ensure they get back to the c2 rust prints wont get recorded). If you want to preserve debug strings like this you can wrap them in an if debug and set debug to false.

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:219]: How common is an error here?
Idk how it works but could see something like loopback error ing out. If this is a common error we may want to handle it and include a "interface errored" status.

This looks good though 👌

**Review by hulto — CHANGES_REQUESTED** [review]
First pass.
Few thoughts on error handling.
Please update the arp scan docs.

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (jabbate19) [reply implants/lib/eldritch/src/pivot/arp_scan_impl.rs:207]: Should never happen

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [inline implants/lib/eldritch/src/pivot/arp_scan_impl.rs:207]: One rascally println left.

Comment (jabbate19) [reply implants/lib/eldritch/src/pivot/arp_scan_impl.rs:207]: How would I do the debug thing?

`static DEBUG: bool = false`

`if DEBUG {}`

?

Comment (jabbate19) [reply implants/lib/eldritch/src/pivot/arp_scan_impl.rs:207]: Or `#[cfg(debug_assertions)]`

**Review by hulto — APPROVED** [review]
There's one println left and please add docs to the `docs/_docs/user-guide/eldritch.md#pivot.arp_scan` currently just has boiler plate.
When adding docs make sure to include an example of the Dict output and document any edge cases or OS requirements.

Once those two are done go ahead and merge o7

File: implants/lib/eldritch/src/pivot/arp_scan_impl.rs
Comment (hulto) [reply implants/lib/eldritch/src/pivot/arp_scan_impl.rs:207]: Both work.
I tend to do static DEBUG: bool but that's more for imix stuff.
cfg(debug_assertions) is perfect.

**Pull request merged**
